src/Grpc.NetCore.HttpClient/Internal/GrpcCall.cs

```diff
@@ -0,0 +1,421 @@
+        {
+            _callCts.Cancel();
+        }
+
+        private void StartSend(System.Net.Http.HttpClient client, HttpRequestMessage message)
+        {
+            SendTask = SendAsync(client, message);
+        }
+
+        private async Task SendAsync(System.Net.Http.HttpClient client, HttpRequestMessage message)
+        {
+            HttpResponse = await client.SendAsync(message, HttpCompletionOption.ResponseHeadersRead, _callCts.Token).ConfigureAwait(false);
+            ValidateHeaders();
+        }
+
+        private HttpContentClientStreamWriter<TRequest, TResponse> CreateWriter(HttpRequestMessage message)
+        {
+            var writeStreamTcs = new TaskCompletionSource<Stream>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var completeTcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            // Canceling call will cancel pending writes to the stream
+            _writerCtsRegistration = _callCts.Token.Register(() =>
+            {
+                completeTcs.TrySetCanceled();
+                writeStreamTcs.TrySetCanceled();
+            });
+
+            message.Content = new PushStreamContent(
+                (stream) =>
+                {
+                    writeStreamTcs.SetResult(stream);
+                    return completeTcs.Task;
+                },
+                GrpcProtocolConstants.GrpcContentTypeHeaderValue);
+
+            var writer = new HttpContentClientStreamWriter<TRequest, TResponse>(this, writeStreamTcs.Task, completeTcs);
+            return writer;
+        }
+
+        private HttpRequestMessage CreateHttpRequestMessage()
+        {
+            var message = new HttpRequestMessage(HttpMethod.Post, Method.FullName);
+            message.Version = new Version(2, 0);
+
+            if (Options.Headers != null && Options.Headers.Count > 0)
+            {
+                foreach (var entry in Options.Headers)
+                {
+                    // Deadline is set via CallOptions.Deadline
+                    if (entry.Key == GrpcProtocolConstants.TimeoutHeader)
+                    {
+                        continue;
+                    }
+
+                    var value = entry.IsBinary ? Convert.ToBase64String(entry.ValueBytes) : entry.Value;
+                    message.Headers.Add(entry.Key, value);
+                }
+            }
+
+            if (_timeout != null)
+            {
+                message.Headers.Add(GrpcProtocolConstants.TimeoutHeader, GrpcProtocolHelpers.EncodeTimeout(Convert.ToInt64(_timeout.Value.TotalMilliseconds)));
+            }
+
+            return message;
+        }
+
+        private void DeadlineExceeded(object state)
+        {
+            if (!_callCts.IsCancellationRequested)
+            {
+                // Flag is used to determin status code when generating exceptions
+                DeadlineReached = true;
+
+                _callCts.Cancel();
```

Comment:
Is there any way user code could block this call to _callCts.Cancel()?

Reply:
I don't believe so. The cancellation token is only exposed internally.